### DISCUSSION TITLE
Sentence Segmentation using senter with parser (Spacy 3.0)

### DISCUSSION THREAD

**Discussion by Phat-Loc** [explosion/spaCy#7903] [Help: Coding & Implementations] [answered]
I am trying to train senter (SentenceRecognizer) and use it with parser (DependencyParser). I have the following questions

1. Is it possible to use both senter and parser? Doesn't the parser do its own sentence segmentation?
2. Can I train a senter to run first in the pipeline then use parser? If so will parser respect the sentence segmentation that senter has set?
3. Given that I have a list of sentence boundaries e.g. [{'start': 0, 'end': 248}, ...] these are character indexes of the doc text which do line up with tokens. What do I set on Doc or Token object so that I can use it training data for senter? Do I set is_sent_start=True for the start of each sentence and false otherwise?

**Comment by polm** [comment]
> Is it possible to use both senter and parser? Doesn't the parser do its own sentence segmentation?

You can use both. In theory the parser respects sentence boundaries set on the Doc before it receives it. 

> Can I train a senter to run first in the pipeline then use parser? If so will parser respect the sentence segmentation that senter has set?

Yes, this is how you'd use them together. The parser should respect the earlier segmentation, and in general it does, though we have seen bugs if `is_sent_start` is set for only some tokens in a doc (#7716). We haven't had any bug reports if you set `is_sent_start` for all tokens, which is how the SentenceRecognizer works and is an expected usage pattern.

> Given that I have a list of sentence boundaries e.g. [{'start': 0, 'end': 248}, ...] these are character indexes of the doc text which do line up with tokens. What do I set on Doc or Token object so that I can use it training data for senter? Do I set is_sent_start=True for the start of each sentence and false otherwise?

Yes, just set `is_sent_start = True` on the first token of each sentence and false otherwise.

**Comment by adrianeboyd** [comment, marked as answer]
A few more details:

Currently in v3.0 the parser is updated in isolation, so it doesn't see the sentence boundaries from the senter during the training update step. During the eval steps on the dev set, the boundaries from the senter do affect the parser, so the performance of the combined components will affect the evaluation and early stopping while training.

This will get more flexible in v3.1, when it will be possible to tell the `senter` to set the boundaries so that the `parser` sees them during training, too. This will train the parser under circumstances that are more similar to the final pipeline, but I'm honestly not sure how it would affect how well/quickly the parser trains in practice, since it may end up making too much of the training data kind of "invalid" because so much of the gold parse tree is unusable if there's a sentence boundary introduced in the middle of a sentence. But if you want to try it out, you can compile the current `develop` branch from source. See the relevant PR: #7767. The basic documentation is in the PR here (not on a published version of the website yet): https://github.com/explosion/spaCy/pull/7767/files#diff-2e35463a512652ee5561262b939fb8bc581b7e113bb36b5853c7c7b5c3219ac2R458

In your training data, you want to set `token.is_sent_start` to either `True` or `False`. You can get the token position of the first token in a span like this:

```python
doc.char_span(start_char, end_char, alignment_mode="expand")[0].i
```

Look at the alignment modes to see what makes sense for your data (in case anything is misaligned, since otherwise you get `None` from `doc.char_span()`): https://spacy.io/api/doc#char_span
